### PR TITLE
SDIT-1378 Add Reconciliation for Basic Incident Detail

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/RetryQueueService.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package uk.gov.justice.digital.hmpps.prisonerfromnomismigration.helpers
+
+import com.fasterxml.jackson.databind.ObjectMapper
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/RetryQueueService.kt:3]: Copied from TO nomis

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
+package uk.gov.justice.digital.hmpps.prisonerfromnomismigration.helpers
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt:1]: Copied from TO Nomis

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt [hunk 3]
```diff
@@ -38,14 +43,21 @@ class IncidentsReconciliationService(
     val dpsClosedIncidentsCount = doApiCallWithRetries { incidentsService.getClosedIncidentsCount(agencyId) }
     val nomisOpenIncidentsCount = nomisIncidents.incidentCount.openIncidents
     val nomisClosedIncidentsCount = nomisIncidents.incidentCount.closedIncidents
+    val openIncidentsMisMatch = checkOpenIncidentsMatch(agencyId, nomisOpenIncidentsCount)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt:46]: Added more detailed check

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationIntTest.kt [hunk 4]
```diff
@@ -29,107 +29,264 @@ class IncidentsReconciliationIntTest : SqsIntegrationTestBase() {
...
       incidentsNomisApi.stubGetIncidentAgencies()
-      incidentsNomisApi.stubGetReconciliationAgencyIncidentCounts("ASI")
-      incidentsNomisApi.stubGetReconciliationAgencyIncidentCounts("BFI")
-      incidentsNomisApi.stubGetReconciliationAgencyIncidentCounts("WWI")
-      incidentsApi.stubGetIncidents()
-    }
 
-    @Test
-    fun `will successfully finish report with no errors`() {
+      incidentsNomisApi.stubGetIncident(33)
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationIntTest.kt:33]: Added separate stub calls to make it easier to understand the calls being made

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt [hunk 5]
```diff
@@ -72,6 +96,89 @@ class IncidentsReconciliationService(
...
+      }
+      .getOrElse { emptyList() }
+      .also { log.info("Page requested: $page, with ${it.size} open incidents") }
+
+  private suspend fun checkOpenIncidentMatch(nomisOpenIncidentId: Long): MismatchOpenIncident? = runCatching {
+    val (nomisOpenIncident, dpsOpenIncident) =
+      withContext(Dispatchers.Unconfined) {
+        async { nomisIncidentsApiService.getIncident(nomisOpenIncidentId) } to
+          async { incidentsService.getIncidentDetailsByNomisId(nomisOpenIncidentId) }
+      }.awaitBoth()
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt:127]: Id suggest you add the doApiCallWithRetries{} wrapper (e.g. see the locations reconciliation) as there can be a lot of network failures.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt (hunk 2 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt:1]: Do you actually need these classes? I cant see where they are used.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt (hunk 5 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/incidents/IncidentsReconciliationService.kt:127]: Good idea - I did that in a few other places.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt (hunk 2 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/helpers/SynchronisationHelpers.kt:1]: I needed SynchronisationHelpers - and then that needed RetryQueueService - I'll shrink down the SynchronisationHelpers for just what I need and remove the other file

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**